tests/unit/utilities/command-generator-test.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+var expect = require('chai').expect;
+var Command = require('../../../lib/utilities/command-generator');
+
+describe('command-generator', function() {
+  it('defaults options', function() {
+    var yarn;
+    yarn = new Command('yarn');
+    expect(yarn.options.networkCommands).to.deep.equal([]);
+
+    yarn = new Command('yarn', {});
+    expect(yarn.options.networkCommands).to.deep.equal([]);
+
+    yarn = new Command('yarn', { networkCommands: ['install'] });
+    expect(yarn.options.networkCommands).to.deep.equal(['install']);
+  });
+
+  it('invoke defaults options', function() {
+    var yarn = new Command('yarn');
+
+    var passedOptions = {};
+    yarn._invoke = function(command, options) {
+      passedOptions = options;
+    };
+
+    // Works with subcommand or argument.
+    yarn.invoke('install');
+    expect(passedOptions).to.deep.equal({ stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke('install', {});
+    expect(passedOptions).to.deep.equal({ stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke('install', { cwd: 'foo' });
+    expect(passedOptions).to.deep.equal({ cwd: 'foo', stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke('install', { stdio: ['default', 'default', 'default'] });
+    expect(passedOptions).to.deep.equal({ stdio: ['default', 'default', 'default'] });
+
+    // Works with no subcommand or argument.
+    yarn.invoke();
+    expect(passedOptions).to.deep.equal({ stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke({});
+    expect(passedOptions).to.deep.equal({ stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke({ cwd: 'foo' });
+    expect(passedOptions).to.deep.equal({ cwd: 'foo', stdio: ['ignore', 'ignore', 'ignore'] });
+
+    yarn.invoke({ stdio: ['default', 'default', 'default'] });
+    expect(passedOptions).to.deep.equal({ stdio: ['default', 'default', 'default'] });
+  });
+
+  it('gets clever on ci', function() {
+    var original;
+    var yarn = new Command('yarn', { networkCommands: ['install'] });
+
+    original = process.env.TRAVIS;
+    process.env.TRAVIS = 'true';
+    expect(yarn.ci('install')).to.deep.equal(['travis_retry']);
+    process.env.TRAVIS = original;
```

Comment:
we should put those in `afterEach()` blocks to make sure the original is restored even if the assertion fails.

```js
describe('in CI environments', function() {
  var originalTravis, originalAppVeyor;

  beforeEach(function() {
    originalTravis = process.env.TRAVIS;
    originalAppVeyor = process.env.APPVEYOR;
  });

  afterEach(function() {
    process.env.TRAVIS = originalTravis;
    process.env.APPVEYOR = originalAppVeyor;
  });

  it ('gets clever on TravisCI', function() {
    process.env.TRAVIS = 'true';

    var yarn = new Command('yarn', { networkCommands: ['install'] });
    expect(yarn.ci('install')).to.deep.equal(['travis_retry']);
  });

  it ('gets clever on AppVeyor', function() {
    process.env.APPVEYOR = 'True';

    var yarn = new Command('yarn', { networkCommands: ['install'] });
    expect(yarn.ci('install')).to.deep.equal(['appveyor-retry']);
  });
});
```